Benchmarks fail to build with C++11

The following benchmarks fail to build if `HPX_WITH_CXX11=ON` using gcc 8.1 and 7.3:

```
benchmark_remove_if
benchmark_unique
benchmark_remove
```

Output is:

```
$ make -kj1 tests.performance.parallel_algorithms.local_
[  0%] Built target hpx_init
[ 90%] Built target hpx
[ 90%] Building CXX object tests/performance/parallel_algorithms/local/CMakeFiles/benchmark_remove_if_exe.dir/benchmark_remove_if.cpp.o
In file included from /nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/bits/char_traits.h:39:0,
                 from /nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/ios:40,
                 from /nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/ostream:38,
                 from include/boost/system/error_code.hpp:19,
                 from /home/simbergm/src/hpx/hpx/error.hpp:15,
                 from /home/simbergm/src/hpx/hpx/exception_fwd.hpp:13,
                 from /home/simbergm/src/hpx/hpx/hpx_finalize.hpp:12,
                 from /home/simbergm/src/hpx/hpx/hpx_init.hpp:15,
                 from /home/simbergm/src/hpx/tests/performance/parallel_algorithms/local/benchmark_remove_if.cpp:8:
/nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/bits/stl_algobase.h: In instantiation of ‘bool std::equal(_IIter1, _IIter1, _IIter2, _BinaryPredicate) [with _IIter1 = __gnu_cxx::__normal_iterator<const int*, std::vector<int> >; _IIter2 = __gnu_cxx::__normal_iterator<const int*, std::vector<int> >; _BinaryPredicate = __gnu_cxx::__normal_iterator<const int*, std::vector<int> >]’:
/home/simbergm/src/hpx/tests/performance/parallel_algorithms/local/benchmark_remove_if.cpp:63:49:   required from here
/nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/bits/stl_algobase.h:1082:25: error: no match for call to ‘(__gnu_cxx::__normal_iterator<const int*, std::vector<int> >) (const int&, const int&)’
  if (!bool(__binary_pred(*__first1, *__first2)))
            ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/bits/stl_algobase.h: In instantiation of ‘bool std::equal(_IIter1, _IIter1, _IIter2, _BinaryPredicate) [with _IIter1 = const int*; _IIter2 = const int*; _BinaryPredicate = const int*]’:
/home/simbergm/src/hpx/tests/performance/parallel_algorithms/local/benchmark_remove_if.cpp:82:49:   required from here
/nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/bits/stl_algobase.h:1082:25: error: ‘__binary_pred’ cannot be used as a function
make[3]: *** [tests/performance/parallel_algorithms/local/CMakeFiles/benchmark_remove_if_exe.dir/build.make:63: tests/performance/parallel_algorithms/local/CMakeFiles/benchmark_remove_if_exe.dir/benchmark_remove_if.cpp.o] Error 1
make[3]: Target 'tests/performance/parallel_algorithms/local/CMakeFiles/benchmark_remove_if_exe.dir/build' not remade because of errors.
make[2]: *** [CMakeFiles/Makefile2:15631: tests/performance/parallel_algorithms/local/CMakeFiles/benchmark_remove_if_exe.dir/all] Error 2
[ 90%] Building CXX object tests/performance/parallel_algorithms/local/CMakeFiles/benchmark_unique_exe.dir/benchmark_unique.cpp.o
In file included from /nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/bits/char_traits.h:39:0,
                 from /nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/ios:40,
                 from /nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/ostream:38,
                 from include/boost/system/error_code.hpp:19,
                 from /home/simbergm/src/hpx/hpx/error.hpp:15,
                 from /home/simbergm/src/hpx/hpx/exception_fwd.hpp:13,
                 from /home/simbergm/src/hpx/hpx/hpx_finalize.hpp:12,
                 from /home/simbergm/src/hpx/hpx/hpx_init.hpp:15,
                 from /home/simbergm/src/hpx/tests/performance/parallel_algorithms/local/benchmark_unique.cpp:8:
/nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/bits/stl_algobase.h: In instantiation of ‘bool std::equal(_IIter1, _IIter1, _IIter2, _BinaryPredicate) [with _IIter1 = __gnu_cxx::__normal_iterator<const int*, std::vector<int> >; _IIter2 = __gnu_cxx::__normal_iterator<const int*, std::vector<int> >; _BinaryPredicate = __gnu_cxx::__normal_iterator<const int*, std::vector<int> >]’:
/home/simbergm/src/hpx/tests/performance/parallel_algorithms/local/benchmark_unique.cpp:63:49:   required from here
/nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/bits/stl_algobase.h:1082:25: error: no match for call to ‘(__gnu_cxx::__normal_iterator<const int*, std::vector<int> >) (const int&, const int&)’
  if (!bool(__binary_pred(*__first1, *__first2)))
            ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/bits/stl_algobase.h: In instantiation of ‘bool std::equal(_IIter1, _IIter1, _IIter2, _BinaryPredicate) [with _IIter1 = const int*; _IIter2 = const int*; _BinaryPredicate = const int*]’:
/home/simbergm/src/hpx/tests/performance/parallel_algorithms/local/benchmark_unique.cpp:82:49:   required from here
/nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/bits/stl_algobase.h:1082:25: error: ‘__binary_pred’ cannot be used as a function
make[3]: *** [tests/performance/parallel_algorithms/local/CMakeFiles/benchmark_unique_exe.dir/build.make:63: tests/performance/parallel_algorithms/local/CMakeFiles/benchmark_unique_exe.dir/benchmark_unique.cpp.o] Error 1
make[3]: Target 'tests/performance/parallel_algorithms/local/CMakeFiles/benchmark_unique_exe.dir/build' not remade because of errors.
make[2]: *** [CMakeFiles/Makefile2:15240: tests/performance/parallel_algorithms/local/CMakeFiles/benchmark_unique_exe.dir/all] Error 2
[ 90%] Built target benchmark_is_heap_until_exe
[ 90%] Built target tests.performance.parallel_algorithms.local_.benchmark_is_heap_until
[ 90%] Built target benchmark_unique_copy_exe
[ 90%] Built target tests.performance.parallel_algorithms.local_.benchmark_unique_copy
[ 90%] Built target benchmark_is_heap_exe
[ 90%] Built target tests.performance.parallel_algorithms.local_.benchmark_is_heap
[ 90%] Built target benchmark_merge_exe
[ 90%] Built target tests.performance.parallel_algorithms.local_.benchmark_merge
[ 90%] Building CXX object tests/performance/parallel_algorithms/local/CMakeFiles/benchmark_remove_exe.dir/benchmark_remove.cpp.o
In file included from /nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/bits/char_traits.h:39:0,
                 from /nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/ios:40,
                 from /nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/ostream:38,
                 from include/boost/system/error_code.hpp:19,
                 from /home/simbergm/src/hpx/hpx/error.hpp:15,
                 from /home/simbergm/src/hpx/hpx/exception_fwd.hpp:13,
                 from /home/simbergm/src/hpx/hpx/hpx_finalize.hpp:12,
                 from /home/simbergm/src/hpx/hpx/hpx_init.hpp:15,
                 from /home/simbergm/src/hpx/tests/performance/parallel_algorithms/local/benchmark_remove.cpp:8:
/nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/bits/stl_algobase.h: In instantiation of ‘bool std::equal(_IIter1, _IIter1, _IIter2, _BinaryPredicate) [with _IIter1 = __gnu_cxx::__normal_iterator<const int*, std::vector<int> >; _IIter2 = __gnu_cxx::__normal_iterator<const int*, std::vector<int> >; _BinaryPredicate = __gnu_cxx::__normal_iterator<const int*, std::vector<int> >]’:
/home/simbergm/src/hpx/tests/performance/parallel_algorithms/local/benchmark_remove.cpp:63:49:   required from here
/nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/bits/stl_algobase.h:1082:25: error: no match for call to ‘(__gnu_cxx::__normal_iterator<const int*, std::vector<int> >) (const int&, const int&)’
  if (!bool(__binary_pred(*__first1, *__first2)))
            ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/bits/stl_algobase.h: In instantiation of ‘bool std::equal(_IIter1, _IIter1, _IIter2, _BinaryPredicate) [with _IIter1 = const int*; _IIter2 = const int*; _BinaryPredicate = const int*]’:
/home/simbergm/src/hpx/tests/performance/parallel_algorithms/local/benchmark_remove.cpp:82:49:   required from here
/nix/store/ng5kbr6kmh979qaachifbsbblydrcmqc-gcc-7.3.0/include/c++/7.3.0/bits/stl_algobase.h:1082:25: error: ‘__binary_pred’ cannot be used as a function
make[3]: *** [tests/performance/parallel_algorithms/local/CMakeFiles/benchmark_remove_exe.dir/build.make:63: tests/performance/parallel_algorithms/local/CMakeFiles/benchmark_remove_exe.dir/benchmark_remove.cpp.o] Error 1
make[3]: Target 'tests/performance/parallel_algorithms/local/CMakeFiles/benchmark_remove_exe.dir/build' not remade because of errors.
make[2]: *** [CMakeFiles/Makefile2:15566: tests/performance/parallel_algorithms/local/CMakeFiles/benchmark_remove_exe.dir/all] Error 2
[100%] Built target benchmark_partition_exe
[100%] Built target tests.performance.parallel_algorithms.local_.benchmark_partition
[100%] Built target benchmark_inplace_merge_exe
[100%] Built target tests.performance.parallel_algorithms.local_.benchmark_inplace_merge
[100%] Built target benchmark_partition_copy_exe
[100%] Built target tests.performance.parallel_algorithms.local_.benchmark_partition_copy
make[2]: Target 'tests/performance/parallel_algorithms/CMakeFiles/tests.performance.parallel_algorithms.local_.dir/all' not remade because of errors.
make[1]: *** [CMakeFiles/Makefile2:14983: tests/performance/parallel_algorithms/CMakeFiles/tests.performance.parallel_algorithms.local_.dir/rule] Error 2
make[1]: Target 'tests.performance.parallel_algorithms.local_' not remade because of errors.
make: *** [Makefile:5128: tests.performance.parallel_algorithms.local_] Error 2
```

--- a/tests/performance/parallel_algorithms/local/benchmark_remove.cpp
+++ b/tests/performance/parallel_algorithms/local/benchmark_remove.cpp
@@ -33,8 +33,8 @@ unsigned int seed = std::random_device{}();
 struct random_fill
 {
     random_fill(std::size_t random_range)
-        : gen(seed),
-          dist(0, random_range - 1)
+      : gen(seed),
+        dist(0, static_cast<int>(random_range - 1))
     {}
 
     int operator()()
@@ -59,8 +59,7 @@ struct vector_type
 
     bool operator==(vector_type const & t) const
     {
-        return std::equal(std::begin(vec_), std::end(vec_),
-            std::begin(t.vec_), std::end(t.vec_));
+        return vec_ == t.vec_;
     }
 
     std::vector<int> vec_;
@@ -78,8 +77,7 @@ struct array_type
 
     bool operator==(array_type const & t) const
     {
-        return std::equal(std::begin(arr_), std::end(arr_),
-            std::begin(t.arr_), std::end(t.arr_));
+        return arr_ == t.arr_;
     }
 
     static const std::size_t arr_size_{ 30 };
@@ -101,7 +99,7 @@ double run_remove_benchmark_std(int test_count,
             org_first, org_last, first);
 
         std::uint64_t elapsed = hpx::util::high_resolution_clock::now();
-        std::remove(first, last, value);
+        (void)std::remove(first, last, value);
         time += hpx::util::high_resolution_clock::now() - elapsed;
     }
 
@@ -153,7 +151,7 @@ void run_benchmark(std::size_t vector_size, int test_count,
         random_fill(random_range));
     org_v = v;
 
-    auto value = DataType(random_range / 2);
+    auto value = DataType(static_cast<int>(random_range / 2));
 
     auto dest_dist = std::distance(first, std::remove(first, last, value));
 

--- a/tests/performance/parallel_algorithms/local/benchmark_remove_if.cpp
+++ b/tests/performance/parallel_algorithms/local/benchmark_remove_if.cpp
@@ -33,8 +33,8 @@ unsigned int seed = std::random_device{}();
 struct random_fill
 {
     random_fill(std::size_t random_range)
-        : gen(seed),
-          dist(0, random_range - 1)
+      : gen(seed),
+        dist(0, static_cast<int>(random_range - 1))
     {}
 
     int operator()()
@@ -59,8 +59,7 @@ struct vector_type
 
     bool operator==(vector_type const & t) const
     {
-        return std::equal(std::begin(vec_), std::end(vec_),
-            std::begin(t.vec_), std::end(t.vec_));
+        return vec_ == t.vec_;
     }
 
     std::vector<int> vec_;
@@ -78,8 +77,7 @@ struct array_type
 
     bool operator==(array_type const & t) const
     {
-        return std::equal(std::begin(arr_), std::end(arr_),
-            std::begin(t.arr_), std::end(t.arr_));
+        return arr_ == t.arr_;
     }
 
     static const std::size_t arr_size_{ 30 };
@@ -152,7 +150,7 @@ void run_benchmark(std::size_t vector_size, int test_count,
         random_fill(random_range));
     org_v = v;
 
-    auto value = DataType(random_range /2);
+    auto value = DataType(static_cast<int>(random_range / 2));
     auto pred = [value](DataType const& a) -> bool { return a == value; };
 
     auto dest_dist = std::distance(first, std::remove_if(first, last, pred));

--- a/tests/performance/parallel_algorithms/local/benchmark_unique.cpp
+++ b/tests/performance/parallel_algorithms/local/benchmark_unique.cpp
@@ -34,7 +34,7 @@ struct random_fill
 {
     random_fill(std::size_t random_range)
         : gen(seed),
-        dist(0, random_range - 1)
+        dist(0, static_cast<int>(random_range - 1))
     {}
 
     int operator()()
@@ -59,8 +59,7 @@ struct vector_type
 
     bool operator==(vector_type const & t) const
     {
-        return std::equal(std::begin(vec_), std::end(vec_),
-            std::begin(t.vec_), std::end(t.vec_));
+        return vec_ == t.vec_;
     }
 
     std::vector<int> vec_;
@@ -78,8 +77,7 @@ struct array_type
 
     bool operator==(array_type const & t) const
     {
-        return std::equal(std::begin(arr_), std::end(arr_),
-            std::begin(t.arr_), std::end(t.arr_));
+        return arr_ == t.arr_;
     }
 
     static const std::size_t arr_size_{ 30 };
@@ -100,7 +98,7 @@ double run_unique_benchmark_std(int test_count,
             org_first, org_last, first);
 
         std::uint64_t elapsed = hpx::util::high_resolution_clock::now();
-        std::unique(first, last);
+        (void)std::unique(first, last);
         time += hpx::util::high_resolution_clock::now() - elapsed;
     }
 

